Make `bias` a ReplicatedTensor in sharded linear.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77307

`bias` was implicitly considered a ReplicatedTensor as part of the
sharded linear implementation.

Since, we now have ReplicatedTensor to express this and also support for DDP,
we can express this as ReplicatedTensor and have our ShardedTensor +
ReplicatedTensor inter-op deal with the associated ops in sharded linear.

Differential Revision: [D36329170](https://our.internmc.facebook.com/intern/diff/D36329170/)

pritamdamania87 committed:
Make `bias` a ReplicatedTensor in sharded linear.

`bias` was implicitly considered a ReplicatedTensor as part of the
sharded linear implementation.

Since, we now have ReplicatedTensor to express this and also support for DDP,
we can express this as ReplicatedTensor and have our ShardedTensor +
ReplicatedTensor inter-op deal with the associated ops in sharded linear.

Differential Revision: [D36329170](https://our.internmc.facebook.com/intern/diff/D36329170/)

[ghstack-poisoned]

wanchaol commented on test/distributed/_shard/sharded_tensor/ops/test_linear.py:
```diff
@@ -125,22 +136,26 @@ def _run_sharded_linear(
+        previous_sharded_bias = sharded_linear.module.bias.clone()
+
+        # Use different learning rate (scale by world_size) for bias since it is replicated via DDP.
+        bias_optim = torch.optim.SGD([sharded_linear.module.bias], lr=0.1 * self.world_size)
```
why we need to use different learning rate instead of a single one? Im not quite sure I can get it, is this also what the end user code look like?

pritamdamania87 replied:
This is covered in our DDP docs: https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel

```
When a model is trained on M nodes with batch=N, the gradient will be M times smaller when compared to the same model trained on a single node with batch=M*N if the loss is summed (NOT averaged as usual) across instances in a batch (because the gradients between different nodes are averaged). You should take this into consideration when you want to obtain a mathematically equivalent training process compared to the local training counterpart. But in most cases, you can just treat a DistributedDataParallel wrapped model, a DataParallel wrapped model and an ordinary model on a single GPU as the same (E.g. using the same learning rate for equivalent batch size).
```

wanchaol replied:
I see, does it mean that we are only doing this (initializing two optimizers) purely for local megatron numerical parity, but for most cases, we don't need to do it with DDP?

fduwjj replied:
Numerically, they are the same. But isn't this going to confuse users because we need to use two learning rates for bias and weight. I am not sure if using separate learning rate is a common practice or not.

pritamdamania87 replied:
I think we will probably have to document this similar to how DDP above documents how to use an appropriate learning rate.

fduwjj replied:
I guess what I want to say here is that, this might still cause confusions to users even if we document it. Hopefully, if we can explain it well in the document so that users won't have too many questions.

wanchaol commented on torch/distributed/_shard/partial_tensor.py:
```diff
@@ -233,6 +234,27 @@ def _transpose_impl(types, args=(), kwargs=None):
+            if lhs.dim() == 1 and rhs.dim() == 1:
+                if lhs.numel() % rhs.numel() != 0:
+                    raise RuntimeError(
+                        f"torch function '{func.__name__}', with args: {args} and "
```
nit: fix lints

pritamdamania87 committed:
Update on "Make `bias` a ReplicatedTensor in sharded linear."


`bias` was implicitly considered a ReplicatedTensor as part of the
sharded linear implementation.

Since, we now have ReplicatedTensor to express this and also support for DDP,
we can express this as ReplicatedTensor and have our ShardedTensor +
ReplicatedTensor inter-op deal with the associated ops in sharded linear.

Differential Revision: [D36329170](https://our.internmc.facebook.com/intern/diff/D36329170/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good!

fduwjj commented on test/distributed/_shard/sharded_tensor/ops/test_linear.py:
```diff
@@ -65,29 +67,24 @@ def _run_sharded_linear(
         shard_parameter(sharded_linear, "weight", spec)
 
+        # Wrap with DDP enabled with ReplicatedTensor
+        with _ddp_replicated_tensor(True):
+            sharded_linear = DDP(sharded_linear)
```
Does this DDP will automatically wrap `bias` as a Replicated Tensor? What if users don't use DDP, does that mean we don't even allow users to use `bias` as a `torch.Tensor`?

pritamdamania87 replied:
It doesn't do it automatically, you need `with _ddp_replicated_tensor(True):`. The tricky thing is that if users don't use DDP, we kinda end up assuming `bias` is still replicated.

I do feel it makes things cleaner if users always use DDP and we recommend users to always combine DDP with sharding and not use sharding alone, because then it results in things like us treating `bias` implicitly as a ReplicatedTensor and I feel its better to have things be explicit.
